book/src/dev/rfcs/XXXX-asynchronous-script-verification.md

```diff
@@ -0,0 +1,231 @@
+that's required to process some request asynchronously is to clone the
+appropriate handle, move it into an async block, and make the call as part of
+the future. (We might want to use Tokio's blocking API for this, but that's a
+side detail).
+
+Because the state service has exclusive access to the sled database, and the
+state service sees a linearized sequence of state requests, we have an easy
+way to opt in to asynchronous database access. We can perform sled operations
+synchronously in the `Service::call`, waiting for them to complete, and be
+sure that all future requests will see the resulting sled state. Or, we can
+perform sled operations asynchronously in the future returned by
+`Service::call`.
+
+If we perform all *writes* synchronously and allow reads to be either
+synchronous or asynchronous, we ensure that writes cannot race each other.
+Asynchronous reads are guaranteed to read at least the state present at the
+time the request was processed, or a later state.
+
+Now, returning to the UTXO lookup problem, we can map out the possible states
+with this restriction in mind. This description assumes that UTXO storage is
+split into disjoint sets, one in-memory (e.g., blocks after the reorg limit)
+and the other in sled (e.g., blocks after the reorg limit). The details of
+this storage are not important for this design, only that the two sets are
+disjoint.
+
+When the state service processes a `Request::AwaitUtxo(OutPoint)` referencing
+some UTXO `u`, there are three disjoint possibilities:
+
+1. `u` is already contained in an in-memory block storage;
+2. `u` is already contained in the sled UTXO set;
+3. `u` is not yet known to the state service.
+
+In case 3, we need to queue `u` and scan all *future* blocks to see whether
+they contain `u`. However, if we have a mechanism to queue `u`, we can
+perform check 2 asynchronously, because restricting to synchronous writes
+means that any async read will return the current or later state. If `u` was
+in the sled UTXO set when the request was processed, the only way that an
+async read would not return `u` is if the UTXO were spent, in which case the
+service is not required to return a response.
+
+This behavior can be encapsulated into a `PendingUtxos`
+structure described below.
+
+```rust
+// sketch
+#[derive(Default, Debug)]
+struct PendingUtxos(HashMap<OutPoint, oneshot::Sender<TransparentOutput>>);
+
+impl PendingUtxos {
+    // adds the outpoint and returns (wrapped) rx end of oneshot
+    // return can be converted to `Service::Future`
+    pub fn queue(&mut self, outpoint: OutPoint) -> impl Future<Output=Result<Response, ...>>;
+
+    // if outpoint is a hashmap key, remove the entry and send output on the channel
+    pub fn respond(&mut self, outpoint: OutPoint, output: TransparentOutput);
+
+    // extracts a list of new outputs from the block and checks them against the
+    // hashmap keys
+    pub fn check_block(&mut self, block: &Block);
```

Comment:
```suggestion
    // Calls `respond` for each (OutPoint, TransparentOutput) pair in `block`
    pub fn respond_block(&mut self, block: &Block);
```

Reply:
I think `check` is better than `respond` because the function is checking the block against the list of pending requests, whereas `respond_block` reads as responding to the block.  It seems OK to have different names for the different functions, since they do different things: the `respond` function is called with specific response data, while `check_block` is called with a block.

Reply:
I'm okay with it not being `respond_block` but I don't think `check_block` is better, it's incredibly generic, "check_block" implies very little about what is happening to me. If anything I'd prefer to not have the helper because I think this will only end up being used one place, the entire function will boil down to just

```rust
block.transactions().iter().for_each(|transaction| {
    let hash = transaction.to_hash();
    transaction
        .outputs()
        .enumerate()
        .for_each(|(index, output)| {
            let outpoint = OutPoint { hash, index };
            self.respond(outpoint, output);
        });
});
```

we could probably add a helper to the `Block` to simplify it further

```rust
block.outpoints().for_each(|(outpoint, output)| self.respond(outpoint.output));
```

I'm voting that we just remove this from the RFC and we can treat the helper as an implementation detail that we hash out during code review on the impl:

```suggestion
```

Reply:
OK, sounds good!